### PR TITLE
[Optimize](Random distribution) Improve the performance of tablet sin…

### LINKED ISSUE #17388: [Optimize](Random distribution) Improve the performance of tablet sink and delta writer of writing blocks
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

The current distribution model for Doris is as follows:

OlapTableSink seperate the original Block into serveral subblocks of each node(BE) by tablets distribution and distributes subblocks to storage engine of backends, then the storage engine will seperate the subblock into multiple tablets channel and each delta writer will handle partial of the block.

This model causes blocks to be split according to tablets, and the splitting process can be a relatively heavy operation. After splitting, the blocks are distributed to different DeltaWriters (Memtables) through RPCs to TabletChannels. The distribution operation on TabletChannels is also a relatively heavy operation. If the distribution property of the table is RANDOM distribution, then we have the opportunity to distribute the blocks according to the complete block during distribution. The advantage of doing so is to reduce memory copying and improve write locality, similar to appending the entire block to the memtable.

**This optimze could save 10% ~ 20% CPU cost of RANDOM distribution table load when enable load_to_single_tablet**
**Whats more, is that we could even write to the local delta writer from OlapTableSink in the single_replica_load mode**

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by eldenmoon**
…k and delta writer of writing blocks

The current distribution model for Doris is as follows:

OlapTableSink seperate the original Block into serveral subblocks of each node(BE) by tablets distribution and distributes subblocks to storage engine of backends, then the storage engine will seperate the subblock into multiple tablets channel and each delta writer will handle partial of the block.

This model causes blocks to be split according to tablets, and the splitting process can be a relatively heavy operation. After splitting, the blocks are distributed to different DeltaWriters (Memtables) through RPCs to TabletChannels. The distribution operation on TabletChannels is also a relatively heavy operation. If the distribution property of the table is RANDOM distribution, then we have the opportunity to distribute the blocks according to the complete block during distribution. The advantage of doing so is to reduce memory copying and improve write locality, similar to appending the entire block to the memtable.

**This optimze could save 10% ~ 20% CPU cost of RANDOM distribution table load when write to single tablet
Whats more, is that we could even write to the local delta writer from OlapTableSink in the single_replica_load mode**
# Proposed changes

Issue Number: close #17388

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/sink/vtablet_sink.cpp [hunk 1]
```diff
@@ -404,10 +409,24 @@ Status VNodeChannel::open_wait() {
     return status;
 }
 
+template Status VNodeChannel::add_block<true>(
```
Comment (yiguolei) [inline be/src/vec/sink/vtablet_sink.cpp:412]: 这种模板不会提升太多的性能，不要加了，编译会很慢。直接 if else吧

Comment (eldenmoon) [reply be/src/vec/sink/vtablet_sink.cpp:412]: OK

**Comment by eldenmoon** [comment]
run p0

File: be/src/vec/sink/vtablet_sink.cpp [hunk 2]
```diff
@@ -1058,13 +1146,34 @@ Status VOlapTableSink::send(RuntimeState* state, vectorized::Block* input_block,
...
+            for (size_t i = 0; i < filter_col.size(); ++i) {
+                filter_data[i] = !_filter_bitmap.Get(i);
+            }
+            vectorized::Block::filter_block_internal(&block, filter_col, block.columns());
+        }
+        VLOG_DEBUG << "send block by append block";
+        RETURN_IF_ERROR(_append_block_to_single_tablet(state, block));
+        input_block->clear();
+        // recalcute tablet each time
+        _partition_to_tablet_map.clear();
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply be/src/vec/sink/vtablet_sink.cpp:1167]: you are right

**Comment by eldenmoon** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.36 seconds
 stream load tsv:          466 seconds loaded 74807831229 Bytes, about 153 MB/s
 stream load json:         40 seconds loaded 2358488459 Bytes, about 56 MB/s
 stream load orc:          76 seconds loaded 1101869774 Bytes, about 13 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230303150743_clickbench_pr_108183.html

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
@caiconghui Sorry to disturb, any suggestion on this PR ?

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run beut

**Comment by eldenmoon** [comment]
run feut

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]
LGTM

**Pull request merged**